File: baeckjoon/9481/9481.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func Confirm(candis [][]bool, ans []int, num, idx, n int) {
	for i := 0; i < n; i++ {
		// 확정된 위치 후보 모두 지우기
		candis[idx][i] = true
		// num 다른 위치들에서 지우기
		candis[i][num-1] = true
	}
	ans[idx] = num
}

// l2r 왼쪽에서 -> 오른쪽 방향으로 커지는가?
func AllConfirm(candis [][]bool, ans []int, l2r bool, n int) {
	num := 1
	switch l2r {
	case true:
		for i := 0; i < n; i++ {
			for j := 0; j < n; j++ {
				candis[i][j] = true
			}
			ans[i] = num
			num++
		}
	case false:
		for i := n - 1; i >= 0; i-- {
			for j := 0; j < n; j++ {
				candis[i][j] = true
			}
			ans[i] = num
			num++
		}
	}
}

// 후보배제 clue 1인 경우는 안 들어옴
func RemoveCandi(candi [][]bool, idx, clue, n int) [][]bool {
	for i := 0; i < clue-1; i++ {
		candi[idx][n-1-i] = true
	}
	return candi
}

func isFinish(ans []int) bool {
	for _, a := range ans {
		if a == 0 {
			return false
		}
	}
	return true
}

// clue를 만족하냐
func isFit(ans []int, n, l, r int) bool {
	max := 0
	cnt_l, cnt_r := 0, 0
	for i := 0; i < n; i++ {
		if max < ans[i] {
			max = ans[i]
			cnt_l++
			if max == n {
				break
			}
		}
	}
	max = 0
	for i := n - 1; i >= 0; i-- {
		if max < ans[i] {
			max = ans[i]
			cnt_r++
			if max == n {
				break
			}
		}
	}
	return (l == cnt_l) && (r == cnt_r)
}

func BackTrack(candi [][]bool, ans []int, idx, n, l, r int) (ret [][]int) {
	// // 후보 수가 제일 적은 위치 찾기
	// for i := 0; i < n; i++ {

	// }
	// 확정 안 된 시작위치 찾기
	for i := idx; i < n; i++ {
		if ans[i] == 0 {
			idx = i
			break
		}
	}
	//# 경우의 수가 적은 부분 선택하고, 후보배제 적극 활용해야함
	for j, can := range candi[idx] {
		if !can {
			candi_tmp := make([][]bool, n)
			copy(candi_tmp, candi)
			for i := 0; i < n; i++ {
				candi_tmp[i] = make([]bool, n)
				copy(candi_tmp[i], candi[i])
			}
			ans_tmp := make([]int, n)
			copy(ans_tmp, ans)

			Confirm(candi_tmp, ans_tmp, j+1, idx, n)
			if isFinish(ans_tmp) {
				if isFit(ans_tmp, n, l, r) {
					ret = append(ret, ans_tmp)
				} else {
					return ret
				}
			} else {
				ret = append(ret, BackTrack(candi_tmp, ans_tmp, idx, n, l, r)...)
			}
		}
	}
	return ret
}

func Solve(n, l, r int) int {
	ans := make([]int, n)
	// 1로 set 되면 불가능한 후보, bool default 값이 false니까 모든 후보 가능
	// cadis[pos][candi]
	candiNotPossible := make([][]bool, n)
	for i := 0; i < n; i++ {
		candiNotPossible[i] = make([]bool, n)
	}
	// 확정지을 수 있는 거 확정 짓기
	switch l {
	case 1:
		Confirm(candiNotPossible, ans, n, 0, n)
	case n:
		AllConfirm(candiNotPossible, ans, true, n)
	}
	switch r {
	case 1:
		Confirm(candiNotPossible, ans, n, n-1, n)
	case n:
		AllConfirm(candiNotPossible, ans, false, n)
	}
	if r+l == n+1 {
		Confirm(candiNotPossible, ans, n, l-1, n)
	}
	// 후보배제
	candiNotPossible = RemoveCandi(candiNotPossible, 0, l, n)
	candiNotPossible = RemoveCandi(candiNotPossible, n-1, r, n)
	// 한 칸만 남은 후보 있으면 확정
	for i := 0; i < n; i++ {
		idx := 0
		sum := 0
		for j := 0; j < n; j++ {
			if !candiNotPossible[i][j] {
				sum++
				idx = j
			}
		}
		// 확정 num == idx + 1
		if sum == 1 {
			Confirm(candiNotPossible, ans, idx+1, i, n)
		}
	}

	candi_tmp := make([][]bool, n)
	copy(candi_tmp, candiNotPossible)
	for i := 0; i < n; i++ {
		candi_tmp[i] = make([]bool, n)
		copy(candi_tmp[i], candiNotPossible[i])
	}
	ans_tmp := make([]int, n)
	copy(ans_tmp, ans)

	ret := BackTrack(candi_tmp, ans_tmp, 0, n, l, r)
	// for _, re := range ret {
	// 	fmt.Println(re)
	// }
	return len(ret) % 1000000007
}

func main() {
	for {
		reader := bufio.NewReader(os.Stdin)
		bytes, _, _ := reader.ReadLine()
		line := string(bytes)
		strs := strings.Fields(line)
		nums := [3]int{}
		for i := 0; i < 3; i++ {
			nums[i], _ = strconv.Atoi(strs[i])
		}
		if nums[0] == nums[1] && nums[1] == nums[2] && nums[2] == 0 {
			break
		}
		fmt.Println(Solve(nums[0], nums[1], nums[2]))
	}
}

File: baeckjoon/9481/9481.c
#include <stdio.h>
#include <stdlib.h>

#define BOOL

// 사이즈 n인 candidate에서 idx에 위치에 num로 답 확정
void confirm(BOOL int **const candidate, int *const answer, int num, int idx, int n)
{
	for (int i=0; i<n; i++)
	{
		// 확정된 위치 후보 모두 지우기
		candidate[idx][i] = 0;
		// 다른 위치들에서 num 지우기
		candidate[i][num-1] = 0;
	}
	answer[idx] = num;
}

// clue가 1이라서 answer확정
// l2r 왼쪽에서 -> 오른쪽 방향으로 커지는가?
void all_confirm(int *const answer, BOOL int l2r, const int n)
{
	int num = 1;
	// 왼쪽 -> 오른쪽
	if (l2r)
		for (int i=0; i<n; i++)
			answer[i] = num++;
	// 오른쪽 -> 왼쪽
	else
		for (int i=n-1; i>=0; i--)
			answer[i] = num++;
}

void clean_up(BOOL int **candidate, int *const answer, const int n)
{
	for (int i=0; i<n; i++)
		free(candidate[i]);
	free(candidate);
	free(answer);
}

// 사이즈 n인 candidate에서 clue가 idx에 있을 때 후보 지움
// clue가 1, n인 경우는 이미 걸려져서 안 들어옴
// n == 4 일 때
// 2->4x // 적어도 2개가 보여야 하니까 4층 건물은 못 옴, clue가 2일 때 그 바로 앞 칸은 4 불가능, 그 다음 칸에 대해서는 모름
// 3->4x,3x // clue가 1일 때 그 바로 앞 칸은 4,3 불가능, 그 다음 칸에 대해서는 모름
// n == 5 일 때
// 2->5x
// 3->5x,4x
// 4->5x,4x,3x
void remove_candidate(BOOL int **const candidate, int idx, int clue, const int n)
{
	for (int i=0; i<clue-1; i++)
		candidate[idx][n-1-i] = 0;
}

BOOL int **copy_candidate(BOOL int **const candidate, const int n)
{
	BOOL int **ret = malloc(sizeof(int*) * n);
	for (int i=0; i<n; i++)
	{
		ret[i] = malloc(sizeof(int) * n);
		for (int j=0; j<n; j++)
			ret[i][j] = candidate[i][j];
	}
	return (ret);
}

int *copy_answer(int *const answer, const int n)
{
	int *ret = malloc(sizeof(int) * n);
	for (int i=0; i<n; i++)
		ret[i] = answer[i];
	return (ret);
}

BOOL int is_finish(int *const answer, const int n)
{
	for (int i=0; i<n; i++)
		if (answer[i] == 0)
			return (0);
	return (1);
}

// clue를 만족하냐
BOOL int is_fit(int *const answer, const int n, int l, int r)
{
	int max = 0;
	int cnt_l = 0, cnt_r = 0;
	for (int i=0; i<n; i++)
	{
		if (max < answer[i])
		{
			max = answer[i];
			cnt_l++;
			if (max == n)
				break;
		}
	}
	max = 0;
	for (int i=n-1; i>=0; i--)
	{
		if (max < answer[i])
		{
			max = answer[i];
			cnt_r++;
			if (max == n)
				break;
		}
	}
	return ((l == cnt_l) && (r == cnt_r));
}

int backtrack(BOOL int **const candidate, int *const answer, int idx, const int n, int l, int r)
{
	//# 후보 수가 제일 적은 위치 찾게 바꿔야 함
	// 확정 안 된 시작위치 찾기
	for (int i=0; i<n; i++)
		if (answer[i] == 0)
		{
			idx = i;
			break;
		}
	int ret = 0;
	//# 경우의 수가 적은 부분 선택하고, 후보배제 적극 활용하도록 바꿔야 함
	for (int j=0; j<n; j++)
	{
		if (candidate[idx][j])
		{
			BOOL int **tmp_candidate = copy_candidate(candidate, n);
			int *tmp_answer = copy_answer(answer, n);

			confirm(tmp_candidate, tmp_answer, j+1, idx, n);
			if (is_finish(tmp_answer, n))
			{
				if (is_fit(tmp_answer, n, l, r))
					ret += 1;
				else
					return (ret);
			}
			else
				ret += backtrack(tmp_candidate, tmp_answer, idx, n, l, r);

			clean_up(tmp_candidate, tmp_answer, n);
		}
	}
	return (ret);
}

int solve(const int n, int l, int r)
{
	int *answer = malloc(sizeof(int) * n);
	for (int i=0; i<n; i++)
		answer[i] = 0;
	// 1로 set되면 가능한 후보, 처음에 다 가능하다고 가정
	// bool : candidate[pos][candi]
	BOOL int **candidate = malloc(sizeof(int*) * n);
	for (int i=0; i<n; i++)
	{
		candidate[i] = malloc(sizeof(int) * n);
		for (int j=0; j<n; j++)
			candidate[i][j] = 1;
	}
	// 확정지을 수 있는 거 확정 짓기
	if (l == 1)
		confirm(candidate, answer, n, 0, n);
	else if (l == n)
	{
		all_confirm(answer, BOOL 1, n);
		clean_up(candidate, answer, n);
		return (1);
	}
	if (r == 1)
		confirm(candidate, answer, n, n-1, n);
	else if (r == n)
	{
		all_confirm(answer, BOOL 0, n);
		clean_up(candidate, answer, n);
		return (1);
	}
	if (r + l == n + 1)
		confirm(candidate, answer, n, l-1, n);
	// 후보배제
	remove_candidate(candidate, 0, l, n);
	remove_candidate(candidate, n-1, r, n);
	// 한 칸만 남은 후보 있으면 확정
	for (int i=0; i<n; i++)
	{
		int idx = 0;
		int sum = 0;
		for (int j=0; j<n; j++)
		{
			if (candidate[i][j])
			{
				sum++;
				idx = j;
			}
		}
		// 확정 num == idx + 1
		if (sum == 1)
			confirm(candidate, answer, idx+1, i, n);
	}

	BOOL int **tmp_candidate = copy_candidate(candidate ,n);
	int *tmp_answer = copy_answer(answer, n);
	int length = backtrack(candidate, answer, 0, n, l, r);
	clean_up(candidate, answer, n);
	clean_up(tmp_candidate, tmp_answer, n);
	return (length);
}

int main(void)
{
	int n, l, r;

	while (1)
	{
		scanf("%d %d %d", &n, &l, &r);
		if (n == l && l == r && r == 0)
			break;
		printf("%d\n", solve(n, l, r));
	}
}
